src/events/tls.py

```diff
@@ -0,0 +1,163 @@
+        """Handler for `certificates_relation_created` event."""
+        if not self.charm.unit.is_leader():
+            return
+
+        self.charm.state.cluster.update({"tls": "enabled", "switching-encryption": "started"})
+
+    def _on_certs_relation_joined(self, event: RelationJoinedEvent) -> None:
+        """Handler for `certificates_relation_joined` event."""
+        if not self.charm.state.cluster.tls:
+            logger.debug(
+                "certificates relation joined - tls not enabled and not switching encryption - deferring"
+            )
+            event.defer()
+            return
+
+        # generate unit private key if not already created by action
+        if not self.charm.state.unit_server.private_key:
+            self.charm.state.unit_server.update(
+                {"private-key": generate_private_key().decode("utf-8")}
+            )
+
+        # generate unit key/truststore password if not already created by action
+        # self.charm.state.unit_server.update(
+        #     {
+        #         "keystore-password": self.charm.state.unit_server.keystore_password
+        #         or self.charm.workload.generate_password(),
+        #         "truststore-password": self.charm.state.unit_server.truststore_password
+        #         or self.charm.workload.generate_password(),
+        #     }
+        # )
+
+        csr = generate_csr(
+            private_key=self.charm.state.unit_server.private_key.encode("utf-8"),
+            subject=self.charm.state.unit_server.host,
+            sans_ip=self.charm.state.unit_server.sans.get("sans_ip", []),
+            sans_dns=self.charm.state.unit_server.sans.get("sans_dns", []),
+        )
+
+        self.charm.state.unit_server.update({"csr": csr.decode("utf-8").strip()})
+
+        self.certificates.request_certificate_creation(certificate_signing_request=csr)
+
+    def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
+        """Handler for `certificates_available` event after provider updates signed certs."""
+        # avoid setting tls files and restarting
+        if event.certificate_signing_request != self.charm.state.unit_server.csr:
+            logger.error("Can't use certificate, found unknown CSR")
+            return
+
+        # if certificate already exists, this event must be new, flag restart
+        if self.charm.state.unit_server.certificate:
+            self.charm.on[f"{self.charm.restart.name}"].acquire_lock.emit()
+
+        self.charm.state.unit_server.update(
+            {"certificate": event.certificate, "ca-cert": event.ca}
+        )
+
+        self.charm.tls_manager.set_private_key()
+        self.charm.tls_manager.set_ca()
+        self.charm.tls_manager.set_certificate()
+        # self.charm.tls_manager.set_truststore()
+        # self.charm.tls_manager.set_p12_keystore()
+
+        self.charm.on[f"{self.charm.restart.name}"].acquire_lock.emit()
```

Comment:
Why asking for the lock twice in the same method?

Reply:
Very good catch, thanks a lot for that.

We shouldn't even have it once.

Removing both.

Setting the cert to the peer relation should trigger a config change and a restart instead.